crates/bevy_ecs/src/bundle.rs

```diff
@@ -166,6 +166,50 @@ impl BundleInfo {
         });
     }
 
+    /// Like write_components, but does not overwrite existing components on the entity
+    /// # Safety
+    /// table row must exist, entity must be valid
+    #[allow(clippy::clippy::too_many_arguments)]
+    #[inline]
+    pub(crate) unsafe fn write_additional_components<T: Bundle>(
```

Comment:
I would prefer if this function used the `bool` pattern that you used in the resources version as well: reducing code duplication would be nice.

That said, this may reduce performance due to branching, and component insertion is much more performance critical. Probably worth using the `ecs_bench_suite` to see if this matters. I'll find that code and link it for you.

Reply:
And here's the link: https://github.com/cart/ecs_bench_suite/tree/bevy-benches

Reply:
Thanks, I'll test out both versions and share the results.

Reply:
I've attached the criterion ouput of the benchmark. The "previous" run is the currently commited version without the extra branch and the "current" run is using a single `write_components` function with the extra branch. I have no idea what the Bevy team's tolerances are for interpreting the results, but the difference between the two benches looks like noise to me (even if criterion doesn't think so. The only two regressions were in `simple_iter/bevy ` and `sparse_fragmented_iter/bevy_foreach`, which I don't think my changes would have affected.

```
simple_iter/bevy        
                        time:   [16.082 us 16.169 us 16.266 us]
                        change: [+1.2329% +1.6582% +2.1681%] (p = 0.00 < 0.05)
                        Performance has regressed.

sparse_fragmented_iter/bevy_foreach
                        time:   [9.5721 ns 9.5902 ns 9.6105 ns]
                        change: [+1.1353% +1.6180% +2.1096%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

<details><summary>Full Criterion Console Output</summary>

```

simple_insert/bevy      time:   [512.04 us 514.32 us 517.42 us]
                        change: [-1.0599% -0.5425% +0.0728%] (p = 0.07 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
Benchmarking simple_insert/bevy_unbatched: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 7.1s, enable flat sampling, or reduce sample count to 50.
simple_insert/bevy_unbatched
                        time:   [1.4189 ms 1.4301 ms 1.4448 ms]
                        change: [-3.2754% -2.0556% -0.7800%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe

simple_iter/bevy        time:   [16.082 us 16.169 us 16.266 us]
                        change: [+1.2329% +1.6582% +2.1681%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe
simple_iter/bevy_sparse time:   [48.867 us 48.907 us 48.957 us]
                        change: [-3.9397% -3.6838% -3.4775%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low mild
  6 (6.00%) high mild
  3 (3.00%) high severe
simple_iter/bevy_foreach
                        time:   [12.470 us 12.484 us 12.500 us]
                        change: [-0.6045% -0.2110% +0.1167%] (p = 0.28 > 0.05)
                        No change in performance detected.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
simple_iter/bevy_sparse_foreach
                        time:   [45.823 us 45.878 us 45.948 us]
                        change: [-0.6064% -0.1626% +0.2285%] (p = 0.46 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) high mild
  5 (5.00%) high severe

fragmented_iter/bevy    time:   [351.20 ns 354.51 ns 357.82 ns]
                        change: [-3.0265% -1.9177% -0.7676%] (p = 0.00 < 0.05)
                        Change within noise threshold.
fragmented_iter/bevy_foreach
                        time:   [281.49 ns 283.85 ns 286.24 ns]
                        change: [-3.5748% -2.6710% -1.7607%] (p = 0.00 < 0.05)
                        Performance has improved.

sparse_fragmented_iter/bevy
                        time:   [10.865 ns 10.887 ns 10.916 ns]
                        change: [-0.5442% -0.0929% +0.3524%] (p = 0.70 > 0.05)
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  9 (9.00%) high severe
sparse_fragmented_iter/bevy_foreach
                        time:   [9.5721 ns 9.5902 ns 9.6105 ns]
                        change: [+1.1353% +1.6180% +2.1096%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 19 outliers among 100 measurements (19.00%)
  3 (3.00%) low mild
  7 (7.00%) high mild
  9 (9.00%) high severe

schedule/bevy           time:   [67.939 us 68.377 us 68.822 us]
                        change: [-1.8476% -0.8390% +0.1306%] (p = 0.11 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

heavy_compute/bevy      time:   [666.37 us 676.65 us 687.36 us]
                        change: [-4.9185% -2.7976% -0.6826%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking add_remove_component/bevy_table: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.9s, enable flat sampling, or reduce sample count to 50.
add_remove_component/bevy_table
                        time:   [1.3447 ms 1.3485 ms 1.3526 ms]
                        change: [+0.7945% +1.2666% +1.7280%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
Benchmarking add_remove_component/bevy_sparse_set: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.5s, enable flat sampling, or reduce sample count to 60.
add_remove_component/bevy_sparse_set
                        time:   [1.0911 ms 1.0955 ms 1.1004 ms]
                        change: [-3.0107% -0.8551% +1.2577%] (p = 0.46 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

add_remove_component_big/bevy_table
                        time:   [2.9684 ms 2.9844 ms 3.0010 ms]
                        change: [-2.8272% -1.9762% -1.1412%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe
Benchmarking add_remove_component_big/bevy_sparse_set: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 6.2s, enable flat sampling, or reduce sample count to 60.
add_remove_component_big/bevy_sparse_set
                        time:   [1.2346 ms 1.2527 ms 1.2762 ms]
                        change: [-2.2371% +2.2181% +6.7912%] (p = 0.37 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe

get_component/bevy      time:   [799.90 us 812.40 us 826.43 us]
                        change: [-2.2862% +0.2488% +2.7980%] (p = 0.87 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
get_component/bevy_system
                        time:   [800.90 us 807.54 us 815.08 us]
                        change: [-0.5879% +0.2857% +1.1703%] (p = 0.52 > 0.05)
                        No change in performance detected.
Found 19 outliers among 100 measurements (19.00%)
  9 (9.00%) high mild
  10 (10.00%) high severe
```
</details>

[BranchlessVsBranchComponentInsertion.zip](https://github.com/bevyengine/bevy/files/6408822/BranchlessVsBranchComponentInsertion.zip)

Reply:
I agree with you: I don't think it should have affected those at all. Perf should only be affected for component adding. I'm pretty firmly in favor of the bool implementation due to reduced code duplication and more clarity so I'd go with that.

@cart you may want to verify these benchmarks before merging though.

Reply:
I've pushed the commit with the extra branch, I decided to use an enum instead of a boolean to make the intent a little clearer. The intent of this:
```rust
pub fn try_insert_resource<T: Component>(&mut self, value: T) {
    let component_id = self.components.get_or_insert_resource_id::<T>();
    // SAFE: component_id just initialized and corresponds to resource of type T
    unsafe { self.insert_resource_with_id(component_id, value, ComponentCollision::Skip) };
}
```
is much clearer than this:
```rust
pub fn try_insert_resource<T: Component>(&mut self, value: T) {
    let component_id = self.components.get_or_insert_resource_id::<T>();
    // SAFE: component_id just initialized and corresponds to resource of type T
    unsafe { self.insert_resource_with_id(component_id, value, false) };
}
```

Reply:
I agree, that's a nice change.